token/services/interop/pledge/wallet.go

```diff
@@ -0,0 +1,236 @@
+	w := token.GetManagementService(sp, opts...).WalletManager().IssuerWallet(id)
+	if w == nil {
+		return nil, errors.Errorf("issuer Wallet needs to be initialized")
+	}
+	return w, nil
+}
+
+type QueryService interface {
+	// TODO: switch to UnspentTokensIteratorBy(id, typ string) (UnspentTokensIterator, error)
+	ListUnspentTokens() (*token2.UnspentTokens, error)
+}
+
+type IssuerWallet struct {
+	wallet       *token.IssuerWallet
+	queryService QueryService
+}
+
+func NewIssuerWallet(sp view.ServiceProvider, wallet *token.IssuerWallet) *IssuerWallet {
+	tmsID := wallet.TMS().ID()
+	net := network.GetInstance(sp, tmsID.Network, tmsID.Channel)
+	if net == nil {
+		logger.Errorf("could not find network [%s]", tmsID)
+		return nil
+	}
+	v, err := net.TokenVault(tmsID.Namespace)
+	if err != nil {
+		logger.Errorf("failed to get vault for [%s]: [%s]", tmsID, err)
+	}
+
+	return &IssuerWallet{
+		wallet:       wallet,
+		queryService: v.QueryEngine(),
+	}
+}
+
+func (w *IssuerWallet) GetPledgedToken(tokenID *token2.ID) (*token2.UnspentToken, *Script, error) {
+	unspentTokens, err := w.queryService.ListUnspentTokens()
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "token selection failed")
+	}
+	return retrievePledgedToken(unspentTokens, tokenID)
+}
+
+type OwnerWallet struct {
+	wallet       *token.OwnerWallet
+	queryService QueryService
+}
+
+// GetWallet returns the wallet whose id is the passed id.
+// If the passed id is empty, GetWallet has the same behaviour of MyWallet.
+// It returns nil, if no wallet is found.
+func GetWallet(sp view.ServiceProvider, id string, opts ...token.ServiceOption) *token.OwnerWallet {
+	w := token.GetManagementService(sp, opts...).WalletManager().OwnerWallet(id)
+	if w == nil {
+		return nil
+	}
+	return w
+}
+
+func Wallet(sp view.ServiceProvider, wallet *token.OwnerWallet) *OwnerWallet {
+	tmsID := wallet.TMS().ID()
+	net := network.GetInstance(sp, tmsID.Network, tmsID.Channel)
+	if net == nil {
+		logger.Errorf("could not find network [%s]", tmsID)
+		return nil
+	}
+	v, err := net.TokenVault(tmsID.Namespace)
+	if err != nil {
+		logger.Errorf("failed to get vault for [%s]: [%s]", tmsID, err)
+	}
+
+	return &OwnerWallet{
+		wallet:       wallet,
+		queryService: v.QueryEngine(),
+	}
+}
+
+func (w *OwnerWallet) GetPledgedToken(tokenID *token2.ID) (*token2.UnspentToken, *Script, error) {
+	unspentTokens, err := w.queryService.ListUnspentTokens()
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "token selection failed")
+	}
+
+	return retrievePledgedToken(unspentTokens, tokenID)
+}
+
+func retrievePledgedToken(unspentTokens *token2.UnspentTokens, tokenID *token2.ID) (*token2.UnspentToken, *Script, error) {
+	logger.Debugf("[%d] unspent tokens found, search [%s]", len(unspentTokens.Tokens), tokenID)
+
+	var res []*token2.UnspentToken
+	var scripts []*Script
+	for _, tok := range unspentTokens.Tokens {
+		if tok.Id.String() == tokenID.String() {
```

Comment:
We cannot have multiple tokens with the same id. So I would expose another method from the query service `GetUnspentToken(id *token2.ID, typ string) (*token2.UnspentToken, error)`. As an interim solution we can have a decorator that just iterates and finds it (though I think a DB query is equally fast to implement).. In any case I would return at the end of this if.